src/trusted/service_runtime/nacl_copy.c

```diff
@@ -51,17 +55,42 @@ int NaClCopyInFromUserZStr(struct NaClApp *nap,
                            size_t         dst_buffer_bytes,
                            uintptr_t      src_usr_addr) {
   uintptr_t src_sys_addr;
-
   CHECK(dst_buffer_bytes > 0);
+
+
   src_sys_addr = NaClUserToSysAddr(nap, src_usr_addr);
   if (kNaClBadAddress == src_sys_addr) {
     dst_buffer[0] = '\0';
     return 0;
   }
+
+  uintptr_t check_addr = src_usr_addr;
+  int bytes_copied = 0;
+
+  /*
+   * We need to handle NULL terminated strings that possibly could overlap VMMap entries,
+   * are not actually NULL terminated, or are longer than the destination buffer.
+   * To do this we check the mapping of the starting address, and copy what we can from that page.
+   * If it overlaps we continue on to check the next page.
+   * Exit conditions are: 1) page is non-valid, 2) we've found a NULL character, 3) we've copied up to the destination length
+   */
+
   NaClCopyTakeLock(nap);
-  strncpy(dst_buffer, (char *) src_sys_addr, dst_buffer_bytes);
+  while (1) {
+    uintptr_t page_end = NaClVmmapCheckAddrMapping( &nap->mem_map, check_addr >> NACL_PAGESHIFT, 1, NACL_ABI_PROT_READ);
+    if (!page_end) break;
+    int page_room = (page_end << NACL_PAGESHIFT) - check_addr;
+    int dst_bytes_remaining = dst_buffer_bytes - bytes_copied;
+    int copy_bytes = page_room < dst_bytes_remaining ? page_room : dst_bytes_remaining;
+    strncpy(dst_buffer + bytes_copied, (char *) src_sys_addr + bytes_copied, copy_bytes);
+    bytes_copied = bytes_copied + copy_bytes;
+    if (strnlen(dst_buffer, bytes_copied) < bytes_copied) break;
```

Comment:
Shouldn't we just get the output of strncpy and subtract the base pointer rather than having to go over all the bytes again with a strnlen?